src/main/host/descriptor/socket/netlink.rs

```diff
@@ -0,0 +1,1215 @@
+                return Err(Errno::ENOPROTOOPT.into());
+            }
+        }
+
+        Ok(())
+    }
+
+    pub fn bind(
+        socket: &Arc<AtomicRefCell<Self>>,
+        addr: Option<&SockaddrStorage>,
+        _net_ns: &NetworkNamespace,
+        rng: impl rand::Rng,
+    ) -> SyscallResult {
+        let socket_ref = &mut *socket.borrow_mut();
+        socket_ref
+            .protocol_state
+            .bind(&mut socket_ref.common, socket, addr, rng)
+    }
+
+    pub fn readv(
+        &mut self,
+        _iovs: &[IoVec],
+        _offset: Option<libc::off_t>,
+        _flags: libc::c_int,
+        _mem: &mut MemoryManager,
+        _cb_queue: &mut CallbackQueue,
+    ) -> Result<libc::ssize_t, SyscallError> {
+        // we could call NetlinkSocket::recvmsg() here, but for now we expect that there are no code
+        // paths that would call NetlinkSocket::readv() since the readv() syscall handler should have
+        // called NetlinkSocket::recvmsg() instead
+        panic!("Called NetlinkSocket::readv() on a netlink socket.");
+    }
+
+    pub fn writev(
+        &mut self,
+        _iovs: &[IoVec],
+        _offset: Option<libc::off_t>,
+        _flags: libc::c_int,
+        _mem: &mut MemoryManager,
+        _cb_queue: &mut CallbackQueue,
+    ) -> Result<libc::ssize_t, SyscallError> {
+        // we could call NetlinkSocket::sendmsg() here, but for now we expect that there are no code
+        // paths that would call NetlinkSocket::writev() since the writev() syscall handler should have
+        // called NetlinkSocket::sendmsg() instead
+        panic!("Called NetlinkSocket::writev() on a netlink socket");
+    }
+
+    pub fn sendmsg(
+        socket: &Arc<AtomicRefCell<Self>>,
+        args: SendmsgArgs,
+        mem: &mut MemoryManager,
+        _net_ns: &NetworkNamespace,
+        _rng: impl rand::Rng,
+        cb_queue: &mut CallbackQueue,
+    ) -> Result<libc::ssize_t, SyscallError> {
+        let socket_ref = &mut *socket.borrow_mut();
+        socket_ref
+            .protocol_state
+            .sendmsg(&mut socket_ref.common, socket, args, mem, cb_queue)
+    }
+
+    pub fn recvmsg(
+        socket: &Arc<AtomicRefCell<Self>>,
+        args: RecvmsgArgs,
+        mem: &mut MemoryManager,
+        cb_queue: &mut CallbackQueue,
+    ) -> Result<RecvmsgReturn, SyscallError> {
+        let socket_ref = &mut *socket.borrow_mut();
+        socket_ref
+            .protocol_state
+            .recvmsg(&mut socket_ref.common, socket, args, mem, cb_queue)
+    }
+
+    pub fn listen(
+        _socket: &Arc<AtomicRefCell<Self>>,
+        _backlog: i32,
+        _net_ns: &NetworkNamespace,
+        _rng: impl rand::Rng,
+        _cb_queue: &mut CallbackQueue,
+    ) -> Result<(), SyscallError> {
+        // We follow the same approach as UnixSocket
+        log::warn!("We do not yet handle listen request on netlink sockets");
```

Comment:
Do we know if some of these syscalls like `listen()` are actually valid for netlink sockets? If we're not sure and we need to figure out what Linux does first (e.g., via fuzz-style testing), adding a comment to that effect would be helpful. (The same goes for the others below.)

Reply:
I think, no, because the only supported socket types of netlink sockets are SOCK_DGRAM and SOCK_RAW (which is the same as SOCK_DGRAM see netlink(7)).

So `listen`, `accept` are not the things, but `connect` is still a valid syscall just (just like in Datagram-oriented protocol like UDP). For `ioctl`, I don't know much about it and I can't find any doc for netlink like tcp(7) or udp(7).